### PR TITLE
2637 create concurrency protocols

### LINKED ISSUE #2637: Create concurrency protocols
## Description

Our plugins should not depend on whether they are run using multiprocessing or multithreading. Currently, they depend on `threading.Event` and `threading.Lock`. The synchronization classes provided by multiprocessing have the same interface as those provided by multithreading. Unfortunately, Python does not provide any base class or interface that can be used as a type hint.

Create two new [protocols](https://peps.python.org/pep-0544/) in `common.concurrency` that can be used for type hinting.

Note: You can probably just copy the definitions from typeshed (https://github.com/python/typeshed/blob/main/stdlib/threading.pyi) and make them protocols

## Tasks
- [ ] Create a protocol for `Lock`
- [ ] Create a protocol for `Event`

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes part of #2637.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/common/concurrency/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from .protocols import Lock, Event
```
Comment (VakarisZ) [inline monkey/common/concurrency/__init__.py:1]: This could also go in types?

Comment (ilija-lazoroski) [reply monkey/common/concurrency/__init__.py:1]: Types has a bunch of things that are general. These on the other hand are concurrency protocols and I think they have a good place.

File: monkey/common/concurrency/protocols.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    def locked(self) -> bool:
+        ...
+
+
+class Event(Protocol):
+    def is_set(self) -> bool:
+        ...
+
+    def isSet(self) -> bool:
```
[the remark this answers is not in the document]
Comment (ilija-lazoroski) [reply monkey/common/concurrency/protocols.py:31]: Probably. We can remove it if we see that it is not supported now.

File: monkey/common/concurrency/__init__.py (hunk 1 above)
Comment (VakarisZ) [reply monkey/common/concurrency/__init__.py:1]: What do you mean by "general"? Why is this not general?

Comment (ilija-lazoroski) [reply monkey/common/concurrency/__init__.py:1]: These are `concurrency` protocols, I reckon we don't want to put them in general types and have one big file with all the types and protocols inside.

File: monkey/common/concurrency/protocols.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ilija-lazoroski) [reply monkey/common/concurrency/protocols.py:31]: It seems that is just an alias that can be removed.

File: monkey/common/types/concurrency.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+class Lock(Protocol):
+    def __enter__(self) -> bool:
+        ...
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
```
Comment (shreyamalviya) [inline monkey/common/types/concurrency.py:15]:
```suggestion
        exception_type: Optional[Type[BaseException]],
        exception_val: Optional[BaseException],
        exception_tb: Optional[TracebackType],
```

**Review by shreyamalviya — APPROVED** [review]

**Review by mssalvatore — CHANGES_REQUESTED** [review]
Why are there changes to common.types in this PR?

PRs should be small and focused.

File: monkey/common/types/concurrency.py (hunk 3 above)
Comment (mssalvatore) [reply monkey/common/types/concurrency.py:15]: The names here are consistent with what's in the python documentation. For that reason, I wouldn't change them

**Pull request merged**